### ISSUE TITLE
Quick Find UI and Production Build

### ISSUE THREAD

**Issue by francisrupert** [USDA-FSA/fsa-design-system#339]
User Interface and a11y requirements for for the Quick Find functionality, as initially prototyped and developed by @bignamehere at #103.

## UI Requirements

Originally authored at https://github.com/USDA-FSA/fsa-design-system/issues/103#issuecomment-627547238 and designed by @francisrupert.

- [x] Default
  - [x] HTML
  - [x] SCSS
- [x] Key entry
  - [x] HTML
  - [x] SCSS

| Default  | Key entry |
|---|---|
| ![quick-find--empty](https://user-images.githubusercontent.com/1165933/81737020-14cbee80-945d-11ea-96f4-6465c4b24c6e.png)  | ![quick-find](https://user-images.githubusercontent.com/1165933/81737028-17c6df00-945d-11ea-9154-324c3fb5e39a.png)  |

## Accessibility Requirements

Modeled upon https://haltersweb.github.io/Accessibility/autocomplete.html. It's a tedious though simple list, mostly a matter of toggle values and managing keypresses. 

The HTML/CSS delivered by @francisrupert sets the stage for the below to achieved, with a few exceptions. 

### Keyboard Interaction
- [x] `KEYPRESS` any character
    - [x] Render results to `ds-quick-find__output`
- [ ] `UP` and `DOWN`
    - [x] Cycle through results. CSS is in place to show focus: [View Screenshot](https://user-images.githubusercontent.com/1165933/82342712-92d94980-99b7-11ea-9011-91aa4cce2f00.png)
    - [x] Skip `ds-quick-find__output-item--section` items
    - [x] `ds-quick-find__input`'s `aria-expanded` attribute toggles between `false` and `true` if at least one result
    - [ ] `aria-activedescendant` attribute value points to currently selected `ds-quick-find__output-item` `ds-quick-find__input`'s (e.g. `quick-find-results__active`) 
    - [x] Toggle focused `ds-quick-find__output-item`'s `aria-selected` attribute between `false` and `true`
    - [x] Won't Fix - ~~Toggle focused `ds-quick-find__output-item`'s `id` value between `quick-find-results__active` and ` ` (empty)~~
- [x] `ESC`
    - [x] Close the menu
    - [x] Retain focus to `ds-quick-find__input`
- [x] `TAB`
    - [x] Select the currently focused auto-suggestion (if arrowed to one)
    - [x] Close the menu
    - [x] Move focus to the next focusable element (`ds-quick-find__clear` in this case).
- [x] `ENTER|RETURN`
    - [x] If a `ds-quick-find__output-item` has *not* been arrowed to, nothing occurs.
    - [x] If a `ds-quick-find__output-item` *has* been arrowed to (and focused), navigate to its `ds-quick-find__output-link` `href`

### Screen-reader Interaction
- [x] When the search field gains focus an initial instruction is announced
    * Achieved via `aria-describedby` attribute
    * Achieved achieved via input's `aria-describedby="quick-find-instructions"`
- [x] After a pause in key entry, count is announced 
    - [x] `<div aria-live="polite" class="fsa-sr-only">7 results</div>` or...
    - [x] ~~`<div aria-live="polite" class="fsa-sr-only">No results found, try again</div>`~~ Edit: "0 results" is accepted
    - [x] `polite` means they are announced after a pause in user activity (as opposed to `assertive` which announces immediately anything changes - overbearing to listener)
- [x] When suggestions are arrowed through they are spoken aloud
    - [x] Achieved when list item gains focus and toggles `aria-selected` property to `true`, announcing its content

### Aria
- [ ] `ds-quick-find__input`
    - [x] `class="ds-quick-find__input"`
    - [x] `aria-describedby="quick-find-instructions"`
    - [x] `aria-owns="quick-find-results"`
    - [x] `aria-expanded="false|true"` 
    - [x] `aria-autocomplete="list"` 
    - [ ] `aria-activedescendant="unique-id"`
- [x] `ds-quick-find__output`
    - [x] `role="listbox"`
- [x] `ds-quick-find__output-item`
    - [x] `role="option"`
    - [x] `aria-selected="false||true"`
- [x] `ds-quick-find__output-item--section`
    - [x] `aria-hidden="true"` (non-focusable, not announced)

**Comment by bignamehere** [comment]
@francisrupert Just an FYI... the ID of an element shouldn't be changed with Javascript, because it should be seen as a UUID for that page. If any other script or code uses that ID for any reason, it would lose the reference and break.

So this Todo should be removed:

- Toggle focused ds-quick-find__output-item's id value between quick-find-results__active and (empty)

**Comment by francisrupert** [comment]
@bignamehere,

That `id` bit was taken from https://haltersweb.github.io/Accessibility/autocomplete.html. I agree it's probably useless and there's almost always a better option than `id`. I'd prefer it gone too.

Here's at least a [screencast](https://usdagcc.sharepoint.com/:v:/r/sites/FBC-UX/XCut/999.%20Dropbox/autocomplete-dom.mp4?csf=1&web=1&e=3m3xcH) (Sharepoint URL) of how the DOM changed onkeypress and moved that `id` around. 

Additionally, that screencast should also be helpful to tracking all the `class` and `aria` attributes dancing around.

**Comment by bignamehere** [comment]
@francisrupert  I have concerns about how the proposed example utilizes the Arrow Keys to move up and down the list, instead of using the Tab key, which naturally navigates from Anchor tag to Anchor tag. The current code push, has a nasty conditional that has to traverse the DOM and check for siblings of parents, and even cousins if a sibling of a parent doesn't exist. By the sound of my fan going on high, and how slowly it performs, I think I need some help with how to traverse the DOM in a better way. The issue with the Up and Down keys is that while we can run code that can check for siblings and cousins, it all has to happen within the cycle of an event.

I might need to break out the events into their own separate code, but I assumed it would be more performant if nested. Either way, it works, but will melt your computer stand, so it needs to be refactored. Any ideas on how to directly access the next Anchor (if it exists) would we helpful.

**Comment by francisrupert** [comment]
@bignamehere,

I'm planning to publish [`1.15.2`](https://github.com/USDA-FSA/fsa-design-system/milestone/19) tomorrow, so I've moved this Quick Find feature from that release (to the Fridge) so we can give it the time and care to do it right. I've yet to test out what you've got so far, and might not until _after_ I've published `1.15.2`.

**Comment by bignamehere** [comment]
@francisrupert the latest updates have greatly improved performance, so please test yourself. I still believe there might be a better way to... "Access the next anchor in the results panel", without trying to traverse the DOM up and down.

**Comment by francisrupert** [comment]
@bignamehere,

Pull down `feature/339` to view `1.15.2`'s `gh-pages` merged in. I've manually managed a few conflicts, Most notably you'll want to view http://localhost:4000/fsa-design-system/sitemap/ and test the autocomplete to make sure the tier 1|2|3 are good.

**Comment by bignamehere** [comment]
@francisrupert can you explain what `aria-activedescendant` is or how it would be used in this example? There are a few others above that we might need to have a chat about, because I am not sure what is needed with them.

**Comment by francisrupert** [comment]
@bignamehere,

That attribute can be seen in action in the screencast shown in my earlier replay at https://github.com/USDA-FSA/fsa-design-system/issues/339#issuecomment-635448214.

That `id` tracked which item is active within the overall component as you arrow down/up. It's effectively the same one as the `aria-selected`. I know, seems redundant, though it can also apply if the _active_ item does not have _focus_.

A bit more at https://www.w3.org/TR/wai-aria/#aria-activedescendant.

**Comment by francisrupert** [comment]
@bignamehere,

I've further evaluated its performance, and think it's pretty solid. I didn't perceive any lag, and watching the Performance and Memory tabs of the Dev Tools Inspector didn't concern me all that much -- though I'm sure you're a better judge of those implications.

Regarding...

> "I still believe there might be a better way to... "Access the next anchor in the results panel", without trying to traverse the DOM up and down."

One thing that did occur to me was to simplify the HTML structure; specifically eliminate the nested structure between `<ul>`, `<li>`, and `<a>`. To wit...

Instead of...

```html
<ul class="ds-quick-find__output" role="listbox">
  <li class="ds-quick-find__output-item ds-quick-find__output-item--section" aria-hidden="true">
    [section]
  </li>
  <li role="option" aria-selected="true" class="ds-quick-find__output-item">
    <a class="ds-quick-find__output-link" href="link.html">Foo</a>
  </li>
  <li role="option" aria-selected="true" class="ds-quick-find__output-item">
    <a class="ds-quick-find__output-link" href="link.html">Foo</a>
  </li>
</ul>
```

...remove the nested structure to...

```html
<div class="ds-quick-find__output" role="listbox">
  <div class="ds-quick-find__output-item ds-quick-find__output-item--section" aria-hidden="true">...</div>
  <a role="option" aria-selected="true"  class="ds-quick-find__output-link" href="link.html">...</a>
  <a role="option" aria-selected="false" class="ds-quick-find__output-link" href="link.html">...</a>
</div>
```

Yeah, less "HTML purity" but I'm not concerned with that since the `role` balances that out. I haven't dug into how you dance around the DOM, but this outta help. Assuming you agree it could help, this would be worth the investment to further optimize its performance.

**Comment by bignamehere** [comment]
@francisrupert I think it currently works pretty smoothly, and while removing the List code would make it much easier to traverse, there are the section headers to contend with.

I am good with it the current code implementation if you are.

**Comment by francisrupert** [comment]
@bignamehere,

Unless you're still targeting more updates, would you agree that it's at least stable enough to release as is? 

If nothing else, we can consider it an MVP state, keep an eye on it, and optimize over next few weeks as needed or as you judge it can be improved?

I'd like to release it alongside other [`1.15.3` milestone items](https://github.com/USDA-FSA/fsa-design-system/milestone/20) (at least the two items assigned to you) either by end of this week (Thursday for me) or (more likely) front half of next week.

**Comment by francisrupert** [comment]
@bignamehere,

Pull down latest `feature/339` branch and note each of my simple changes.

* 277eafdbb4cfebf57420684f05b5df1466934024 CSS update. Show component only if modern browser

* e9c6c856f91a350f4b74ba50cd449be6db36f69b (nothing to see here)

* 5625292d13f81607cb2123d86819b08f36332867 Hitting SPACE on focused item goes to URL. HItting TAB on focused item does NOT go to URL.

* 6ea5e50b17cc43f08e960bc2cd90288178d0e9a5 Wrapped entire `ds-quick-find.js` contents with a Feature Detection (JS-equivalent of 277eafdbb4cfebf57420684f05b5df1466934024 above.

**Comment by bignamehere** [comment]
@francisrupert Yeah, other than some of the aria stuff that I was learning about today, I think we can launch as is without any new new updates. I will push my latest.

**Comment by francisrupert** [comment]
@bignamehere,

I'm closing this issue as we both agree it's ready to launch. This is just plain awesome, thanks for doing right by it!

Meanwhile, feel free to log an issue for any additional work it still needs, or even might need for a future maintenance release or something.
